### PR TITLE
feat: Catalogue sitemap

### PR DESCRIPTION
**Pull request by connoratrug**
- Generates catalogue resource sitemap on request (/:schema/sitemap.xml)

### CODE REVIEW

**Comment by mswertz** [comment]
a smoke test such as for other APIs in https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java would be nice.

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/controllers/SiteMapController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.molgenis.emx2.MolgenisException;
+import org.molgenis.emx2.Row;
+import org.molgenis.emx2.Schema;
+import org.molgenis.emx2.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+
+public class SiteMapController {
```
Comment (mswertz) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/controllers/SiteMapController.java:24]: I would prefer if this follows same style and naming as https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/JsonYamlApi.java with a create option

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java [hunk 2]
```diff
@@ -55,6 +56,8 @@ public static void start(int port) {
     get(("/" + OIDC_LOGIN_PATH), oidcController::handleLoginRequest);
     get("/" + ROBOTS_TXT, MolgenisWebservice::robotsDotTxt);
 
+    get("/:schema/sitemap.xml", SiteMapController::getSiteMapForSchema);
```
Comment (mswertz) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:59]: Can you please use same patter as all other *Api classes, and use a 'create' function to create these things?

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java [hunk 3]
```diff
@@ -101,14 +104,17 @@ public static void start(int port) {
...
             res.redirect("/" + req.params(SCHEMA) + "/");
           }
         });
     before(
         "/:schema/:app",
         (req, res) -> {
-          if (!req.params("app").equals("graphql") && !req.params("app").equals("theme.css")) {
+          if (!req.params("app").equals("graphql")
+              && !req.params("app").equals("theme.css")
+              && !req.params("app").equals("sitemap.xml")) {
```
Comment (mswertz) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:117]: I would prefer that you move these to the 'api' prefix. Instead of escaping everywhere. The graphql and theme.css are exceptions on that rule for good reasons.

**Review by mswertz — CHANGES_REQUESTED** [review]

**Review by mswertz — CHANGES_REQUESTED** [review]
a smoke test such as for other APIs in https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java would be nice.

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java (hunk 2 above)
Comment (connoratrug) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:59]: why the extra indirection , this is a class that has only a single function / method , it is a static class with a static method. Maybe discuss this offline , i would prefer non static singleton classes that have there deps injected for easy testing

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java (hunk 3 above)
Comment (connoratrug) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:117]: i would prefer to remove the before handlers that change the path these are to greedy

**Comment by connoratrug** [comment]
> a smoke test such as for other APIs in https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java would be nice.

yes, but this is missing the catalogue setup , we talked about having a pet-store catalogue but i am not up to date in the status of that , i think that may have morphed into a other story

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java (hunk 2 above)
Comment (mswertz) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:59]: two considerations: (1) consistency with what is already there, so if we change this we do it everywhere and (2) reason why we use 'create()' is such that all code related to this api is in same place. Because there is a tight coupling between these two pieces of code because of the variable mapping.

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java (hunk 3 above)
Comment (mswertz) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisWebservice.java:117]: they might not be necessary if we initialise handlers in correct order.

**Comment by mswertz** [comment]
> > a smoke test such as for other APIs in https://github.com/molgenis/molgenis-emx2/blob/master/backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java would be nice.
> 
> yes, but this is missing the catalogue setup , we talked about having a pet-store catalogue but i am not up to date in the status of that , i think that may have morphed into a other story

I don't want to preload catalogue such that all emx2 instances are burdened with it (e.g. rare disease team would then need to delete it every time). So instead I recommend you load it for this test following the example of https://github.com/molgenis/molgenis-emx2/blob/816ce0681fb4613367b313da21196b14fc7fc014/backend/molgenis-emx2-io/src/test/java/org/molgenis/emx2/io/TestCohortCatalogue.java#L38

File: backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java [hunk 4]
```diff
@@ -457,6 +458,11 @@ public void testMolgenisWebservice_robotsDotTxt() {
     when().get("/robots.txt").then().statusCode(200).body(equalTo("User-agent: *\nAllow: /"));
   }
 
+  @Test
+  public void testCatalogueSiteMap() {
+    CatalogueExample.create(db);
```
Comment (connoratrug) [inline backend/molgenis-emx2-webapi/src/test/java/org.molgenis.emx2.web/WebApiSmokeTests.java:463]: fails, due to data issue ?

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/controllers/SiteMapController.java (hunk 1 above)
Comment (connoratrug) [reply backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/controllers/SiteMapController.java:10]: combining a 'create' function with a static class does not make any sense to be , JsonYamlApi.create function seams to add certain routes to the spark route request handler, as the order in which the routes added matter hiding these adds in a create function makes it hard to follow the request flow.

**Comment by mswertz** [comment]
refactored to follow standard pattern. I am happy to change it at some point to put all mapping in central place but then with some more nesting.

**Review by mswertz — APPROVED** [review]

**Comment by connoratrug** [comment]
would argue having the hardcoded route in the get is clearer ( especially with the spark-java way of setting the  routes)

**Comment by mswertz** [comment]
> would argue having the hardcoded route in the get is clearer ( especially with the spark-java way of setting the routes)

Yes, I think we should rework all the router properly. Ideally using path groups
https://sparkjava.com/documentation#path-groups
